### ISSUE TITLE
Wheel building with CPU optimization flag error

### ISSUE THREAD

**Issue by h6197627** [numpy/numpy#18892]
Building NumPy python wheel reports error in the end of process if CPU instruction set baseline flag is passed:

> python3 setup.py build --cpu-baseline="avx" bdist_wheel

> AttributeError: 'CCompilerOpt' object has no attribute '_requested_baseline'

Full error:
```
########### EXT COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'avx'
  Enabled     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : 
              : 
  AVX2        : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2
  Extra checks: none
  Detect      : AVX F16C AVX2
              : numpy/core/src/umath/_umath_tests.dispatch.c
              : 
  (FMA3 AVX2) : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2
  Extra checks: none
  Detect      : AVX F16C FMA3 AVX2
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512F     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f
  Extra checks: AVX512F_REDUCE
  Detect      : AVX512F
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512_SKX  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq
  Extra checks: AVX512BW_MASK AVX512DQ_MASK
  Detect      : AVX512_SKX
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_ext.py

########### EXT COMPILER OPTIMIZATION ###########
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/user/Work/numpy-1.20.2/numpy/distutils/command/build_ext.py", line 166, in report
    log.info(self.compiler_opt.report(full=True))
  File "/home/user/Work/numpy-1.20.2/numpy/distutils/ccompiler_opt.py", line 2331, in report
    baseline_rows.append(("Requested", repr(self._requested_baseline)))
AttributeError: 'CCompilerOpt' object has no attribute '_requested_baseline'
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_ext.py

########### CLIB COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'avx'
  Enabled     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_clib.py
```
Although wheel file is successfully created and after installation it passes all standard tests.

Simply building NumPy without wheel goes without errors:
> python3 setup.py build --cpu-baseline="avx"
```
########### EXT COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'avx'
  Enabled     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : 
              : 
  AVX2        : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2
  Extra checks: none
  Detect      : AVX F16C AVX2
              : numpy/core/src/umath/_umath_tests.dispatch.c
              : 
  (FMA3 AVX2) : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2
  Extra checks: none
  Detect      : AVX F16C FMA3 AVX2
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512F     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f
  Extra checks: AVX512F_REDUCE
  Detect      : AVX512F
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512_SKX  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq
  Extra checks: AVX512BW_MASK AVX512DQ_MASK
  Detect      : AVX512_SKX
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_ext.py

########### CLIB COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'avx'
  Enabled     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_clib.py
```
But if after building additionally request to create wheel distribution with the first command (**python3 setup.py build --cpu-baseline="avx" bdist_wheel**) it will end up with the same error.

Also, building wheel without specifying CPU instruction set has no errors:
> python3 setup.py bdist_wheel
```
########### EXT COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : -msse -msse2 -msse3
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : 
              : 
  SSE41       : SSE SSE2 SSE3 SSSE3
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41
              : numpy/core/src/umath/_umath_tests.dispatch.c
              : 
  SSE42       : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX2        : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2
  Extra checks: none
  Detect      : AVX F16C AVX2
              : numpy/core/src/umath/_umath_tests.dispatch.c
              : 
  (FMA3 AVX2) : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2
  Extra checks: none
  Detect      : AVX F16C FMA3 AVX2
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512F     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f
  Extra checks: AVX512F_REDUCE
  Detect      : AVX512F
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
              : 
  AVX512_SKX  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD
  Flags       : -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq
  Extra checks: AVX512BW_MASK AVX512DQ_MASK
  Detect      : AVX512_SKX
              : build/src.linux-x86_64-3.8/numpy/core/src/_simd/_simd.dispatch.c
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_ext.py

########### CLIB COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : -msse -msse2 -msse3
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
CCompilerOpt._cache_write[796] : write cache to path -> /home/user/Work/numpy-1.20.2/build/temp.linux-x86_64-3.8/ccompiler_opt_cache_clib.py
```

### NumPy/Python version information:
1.20.2 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0]

**Comment by seiko2plus** [comment]
This issue has two sides:
1. This fatal error is caused by calling an object whose attributes have been freed.

2. But if you look closely, you will find that the main cause of the issue is duplicate building the extensions. in other words, command `build_ext` has been calling twice one by `build` and the other by `bdist_wheel`. which leads to increasing the build time and also leads to discovering the first side.

the first part of this issue should be fixed by #18899, while still looking for a better way to fix the second one. For a workaround, specify the option `build` after `bdist_wheel` and the two sides of this bug will disappear.

**Comment by seiko2plus** [comment]
@rgommers, Is this a normal behavior to get duplicated builds when option `build` specified before `bdist_wheel`?

**Comment by rgommers** [comment]
Yes, setuptools isn't very smart with reusing; if you give it multiple commands it will execute them separately. If you want to build a wheel, you just do `python setup.py bdist_wheel`.

**Comment by seiko2plus** [comment]
@rgommers, so there's no need to create a workaround for such behavior, and any specified `build` options should come after `bdist_wheel` to avoid any duplicate building; For example, it should be `python setup.py bdist_wheel build --cpu-baseline="avx2"` instead of `python setup.py build --cpu-baseline="avx2" bdist_wheel`, Is that correct? if the answer is yes then we can close this issue since #18899 already fixed the reported error.

**Comment by rgommers** [comment]
I'm a little surprised that that makes a difference, but the whole "multiple commands on a single setup.py invocation" is a little weird anyway. There's an example at http://scipy.github.io/devdocs/building/linux.html for `--compiler=intel`, but other than the rule that flags go _after_ the command I don't think there's many rules/guarantees.

Does `python setup.py bdist_wheel build --cpu-baseline="avx2"` work? I just searched for `bdist_wheel` and found this in `tools/travis-test.sh`:
```
 $PYTHON setup.py build --warn-error build_src --verbose-cfg bdist_wheel
```
and that seems to work (example log [here](https://travis-ci.com/github/numpy/numpy/jobs/502788069)) and not do a double build. So maybe something else is off here.

> Yes, setuptools isn't very smart with reusing; if you give it multiple commands it will execute them separately. 

So this may have been incorrect, it's _supposed_ to work but is badly tested.

**Comment by bashtage** [comment]
I ran into the same when writing the clang-cl patch. In the end, I added a not great `hasattr` check.

https://github.com/numpy/numpy/pull/13816/files#diff-a9126c03aa7a34e7b791e5d0202715b359bb3a44da09d0dbc8f156932cb42240R2348-R2351

**Comment by h6197627** [comment]
> Does python setup.py bdist_wheel build --cpu-baseline="avx2" work?

Yes, changing the order helps to build wheel without errors

> "multiple commands on a single setup.py invocation" is a little weird

I haven't found how to build wheel with instruction set parameter without merging both **build** and **bdist_wheel** into one line
1. `python setup.py bdist_wheel --cpu-baseline="avx2"`
complains about unknown parameter
2. `python setup.py build --cpu-baseline="avx2"`
`python setup.py bdist_wheel`
restarts build from scratch without flag

**Comment by bashtage** [comment]
What happens if you use config first?

**Comment by h6197627** [comment]
I tried
`python3 setup.py config --cpu-baseline="avx" bdist_wheel`
it throws
> error: option --cpu-baseline not recognized

**Comment by seiko2plus** [comment]
My bad (^_^). It seems normal to expect that any distutils command option can be called more than once within the same thread/process, and by looking deep into my code I found that `CCompilerOpt` didn't have the chance to update the cache within multiple calls of method `run()`. #18911 should fix this issue.

@rgommers,
> Does python setup.py bdist_wheel build --cpu-baseline="avx2" work? 

yes, should be the righteous way to avoid duplicate calls of method `build_ext::run()`

@bashtage,
> I ran into the same when writing the clang-cl patch. In the end, I added a not great hasattr check.

Could you please remove this check? this error already fixed by #18899

@h6197627,
> error: option --cpu-baseline not recognized

Its only available through `build`, `build_ext` and `build_clib`.

**Comment by bashtage** [comment]
@seiko2plus I already removed it -- it wasn't necessary once I realized that the build and bdist_wheel command ordered mattered.
